### PR TITLE
BUG: fix `Series.argsort` 

### LINKED ISSUE #12694: BUG: Series.argsort() incorrect handling of NaNs
It appears that `Series.argsort()` is implemented as `s.dropna().argsort().reindex(s.index, fill_value=-1) = np.argsort(s.dropna()).reindex(s.index, fill_value=-1)`.

There are two problems with this:
(a) Since the result represents integer indices into the original series `s`, the result should not have the same `index` as `s.index` -- it should either be a `Series` with index `[0, 1, ...]`, or more likely simply be a NumPy`array`;
(b) The way `NaN`s are effectively removed before calling `np.argsort()` leads to indexes that are no longer appropriate into the original `Series`, resulting in the nonsensical results shown in `[9]` and `[10]` below.

Note that: "As of NumPy 1.4.0 `argsort` works with real/complex arrays containing nan values." So it's not obvious to me that there's much to be gained from mucking around with `np.argsort(s.values)`.

```
Python 3.4.4 (v3.4.4:737efcadf5a6, Dec 20 2015, 20:20:57) [MSC v.1600 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython 4.1.2 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import numpy as np

In [2]: import pandas as pd

In [3]: s = pd.Series([200, 100, 400, 500, np.nan, 300], index=list('abcdef'))

In [4]: s
Out[4]:
a    200.0
b    100.0
c    400.0
d    500.0
e      NaN
f    300.0
dtype: float64

In [5]: s.argsort()
Out[5]:
a    1
b    0
c    4
d    2
e   -1
f    3
dtype: int64

In [6]: s.dropna().argsort().reindex(s.index, fill_value=-1)
Out[6]:
a    1
b    0
c    4
d    2
e   -1
f    3
dtype: int64

In [33]: np.argsort(s.dropna()).reindex(s.index, fill_value=-1)
Out[33]:
a    1
b    0
c    4
d    2
e   -1
f    3
dtype: int64

In [34]: np.argsort(s.dropna().values)
Out[34]: array([1, 0, 4, 2, 3], dtype=int64)

In [7]: np.argsort(s.values)
Out[7]: array([1, 0, 5, 2, 3, 4], dtype=int64)

In [8]: s[np.argsort(s.values)]  # desired result
Out[8]:
b    100.0
a    200.0
f    300.0
c    400.0
d    500.0
e      NaN
dtype: float64

In [9]: s[s.argsort()]  # nonsensical result
Out[9]:
b    100.0
a    200.0
e      NaN
c    400.0
f    300.0
d    500.0
dtype: float64

In [10]: s[s.argsort().values]  # nonsensical result
Out[10]:
b    100.0
a    200.0
e      NaN
c    400.0
f    300.0
d    500.0
dtype: float64

In [11]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.4.4.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 62 Stepping 4, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.1
setuptools: 20.3.1
Cython: None
numpy: 1.10.4
scipy: 0.17.0
statsmodels: 0.6.1
xarray: None
IPython: 4.1.2
sphinx: None
patsy: 0.4.1
dateutil: 2.5.1
pytz: 2016.2
blosc: None
bottleneck: None
tables: 3.2.2
numexpr: 2.5
matplotlib: 1.5.1
openpyxl: 2.3.4
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.4.1
html5lib: 1.0b8
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
jinja2: 2.8
boto: None
```

### LINKED ISSUE #42056: BUG: pd.Series.argsort misaligned index
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

Code:
```python
series = pd.Series([30, 10, 20], index=['c', 'a', 'b'])
order = series.argsort()
order
```
Output:
```
c    1
a    2
b    0
dtype: int64
```

#### Problem description

It appears that the argsort output str index order (c, a, b) remains the same as the input data, while the numeric indices (1, 2, 0) are referring to the input data in a different order. This means the str index no longer corresponds with the values (the numeric indices) in the resulting Series. This behavior is confusing, since it sort of looks like c would be the 1st element (0-indexed), a the 2nd, and b the 0th, which is incorrect.

There's actually another example of this here: https://www.geeksforgeeks.org/python-pandas-series-argsort/. The str indices in the result of the argsort don't give any useful information - only the order of the numeric indices is relevant.

#### Expected Output

I'm not sure if it would make more sense if str indices were reordered to match the numeric indices, like this:
```
a    1
b    2
c    0
dtype: int64
```
This might be nice, for example, to get the indices of the largest N elements: `order.index[-N:]`. The common workflow `series[series.argsort()]` still seems to work, but I'm not sure of all the other indexing implications.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2cb96529396d93b46abab7bbc73a208e708c642e
python           : 3.7.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 3.10.0-1160.25.1.el7.x86_64
Version          : #1 SMP Tue Apr 13 18:55:45 EDT 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.4
numpy            : 1.18.5
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.1
setuptools       : 56.0.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 7.22.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.1
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.3
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by attack68**
- [x] closes #42056
- [x] closes #12694

rendered whatsnew:
Screen Shot 2021-09-18 at 13 56 52

### CODE REVIEW

**Comment by mzeitlin11** [comment]
Would this handle #12694?

**Comment by attack68** [comment]
> Would this handle #12694?

Yes this fixes that.

It solves it by:

 - keeping the '-1' for null values (so works for all dtypes), but doing the indexing correctly.
 - sorting the index for the return series so that there is meaning to its presence.

**Comment by attack68** [comment]
@jorisvandenbossche since you commented on the underlying issue perhaps you would also like to review?

File: pandas/core/series.py [hunk 1]
```diff
@@ -3659,21 +3659,62 @@ def argsort(self, axis=0, kind="quicksort", order=None) -> Series:
...
+        Name: xy, dtype: int64
+        """
+        values = self.values
+        na_mask = isna(values)
+        if not any(na_mask):
+            res = np.argsort(values, kind=kind)
+            res_ser = Series(res, index=self.index[res], dtype="int64", name=self.name)
+            return res_ser.__finalize__(self, method="argsort")
+        else:
+            # GH 42090
```
Comment (jreback) [inline pandas/core/series.py:3693]: would be very clear that we are ordering nulls at the `na_position='last'` (equiv of sort_values). i think its fine not to expose this as an option (as numpy doesn't).

File: pandas/core/series.py [hunk 2]
```diff
@@ -3659,21 +3659,62 @@ def argsort(self, axis=0, kind="quicksort", order=None) -> Series:
         See Also
```
Comment (jreback) [inline pandas/core/series.py:3659]: add See Also to idxmax/min

**Review by jreback — CHANGES_REQUESTED** [review]
haven't really looked, just some quick comments

**Comment by jorisvandenbossche** [comment]
@attack68 thanks for looking into this topic!

Took a quick look, and some high-level comments:

- I agree that the current behaviour of the index doesn't seem very useful (and is confusing), but we shouldn't regard this as a "bug fix" if we completely change the output's index (that's a breaking change)
  - But further thinking about it, is the new behaviour actually ever useful? When would you want to use this index? If we change the behaviour of this method, we could maybe also consider returning an array instead of a Series (and then we don't have this question about what the index should be)
- For the NaN issue, you are now keeping the -1, but another option to fix this part could be to actually give this the proper index value that would sort the series correctly? (as numpy does)

**Comment by attack68** [comment]
> * I agree that the current behaviour of the index doesn't seem very useful (and is confusing), but we shouldn't regard this as a "bug fix" if we completely change the output's index (that's a breaking change)

I interpret a "bug-fix" as a breaking change that corrects erroneous behaviour, and as pointed out by others this function is currently unusable because of its errors:
 - one, the index values returned are wrong if interpreted as being the integer indexes over a Series, which contains Nans.
 - two, since values are re-ordered and the index is not re-ordered any index-value relationship  is broken and made completely meaningless - worse that if anyone interprets it as having meaning will lead to bugs in their program.

> But further thinking about it, is the new behaviour actually ever useful? When would you want to use this index? If we change the behaviour of this method, we could maybe also consider returning an array instead of a Series (and then we don't have this question about what the index should be)

My use of numpy argsort has has been to sort an array by the values of another similar shape array, which is just a kind of key-sorting: `array2[array1.argsort()]`, which was useful in a nearest neighbour algorithm as I recall.

Since numpy arrays don't have index objects per se this acts as an index sort. Since pandas data structures have index objects you can perform the same kind of method on objects with shared indexes with: `ser2[ser1.sort_values().index]`.
So argsort is not really necessary.

But, there are often different ways of doing things. And someone coming from numpy used to using `argsort` for this type of thing might instinctively look for same functionality. In this PR the following is now possible:

`ser2.iloc[ser1.argsort()]`  or `ser2[ser.argsort().index]`

I have added "first" and "last" to ``na_position`` so this does all now.

File: pandas/core/series.py (hunk 2 above)
Comment (attack68) [reply pandas/core/series.py:3664]: added

**Comment by attack68** [comment]
@jreback @jorisvandenbossche this has been updated. Now has the following user defined action:

```python
s1 = pd.Series([30, 10, np.nan, 20, 25], index=["high", "low", "null", "mid-low", "mid-high"])
s1.argsort(na_position=na_position)
```

`na_position` | ... | ...
--- | --- | --- 
`None` | `"first"` | `"last"`
low 1<br>mid-low 3<br>mid-high 4<br>high 0<br>null -1 | null 2<br>low 1<br>mid-low 3<br>mid-high 4<br>high 0 | low 1<br>mid-low 3<br>mid-high 4<br>high 0<br>null 2

**Comment by jreback** [comment]
I am not sure we can actually make this change. The problem is these are meant to mirror `np.argsort` which gives the 'correct' results here.

in your last example can you show what `np.argsort` does to the values?

**Comment by attack68** [comment]
> I am not sure we can actually make this change. The problem is these are meant to mirror `np.argsort` which gives the 'correct' results here.
> 
> in your last example can you show what `np.argsort` does to the values?

The current implementation not only **doesn't mirror** `np.argsort`, but returns incorrect results, confuses the issue further by having an index which is meaningless, and a structure that is useless (having '-1' midway through the results values).

### Current Implementation

```python
>>> s1 = pd.Series([30, 10, np.nan, 20, 25], index=["high", "low", "null", "mid-low", "mid-high"])
>>> s1.argsort()
high        1
low         2
null       -1
mid-low     3
mid-high    0
dtype: int64
>>> np.argsort(s1.values)
array([1, 3, 4, 0, 2])
```

### This PR

My changes here can mirror `np.argsort` (with `na_position="last"`), works with all pandas null values and `dtypes`, has the added benefit of a useable index. `na_position=None` provides a somewhat recognisable `pandas` format:

```python
>>> s1.argsort(na_position="last")
low         1
mid-low     3
mid-high    4
high        0
null        2
dtype: int64
```

**Comment by jreback** [comment]
@attack68 my point is that even though this is a bug, it is likely relied upon. So in an ideal situation we have `np.argsort(..)` and `.argsort` do the same thing. This change might be too big for a simple change and requires deprecation / hard break.

**Comment by jreback** [comment]
`.argsort` was changed relatively recently, can you link the issues and summarize what the problem / fix was?

**Comment by attack68** [comment]
> @attack68 my point is that even though this is a bug, it is likely relied upon. So in an ideal situation we have `np.argsort(..)` and `.argsort` do the same thing. This change might be too big for a simple change and requires deprecation / hard break.

OK, I understand, but by the same token keeping the function as is for the sake of reliability is a valid argument with existing users who have managed to work around it, but invalid with respect to new users of the function for whom it will be unreliable, e.g. the original poster of the underlying issue (not me).

**Comment by jreback** [comment]
> > @attack68 my point is that even though this is a bug, it is likely relied upon. So in an ideal situation we have `np.argsort(..)` and `.argsort` do the same thing. This change might be too big for a simple change and requires deprecation / hard break.
> 
> OK, I understand, but by the same token keeping the function as is for the sake of reliability is a valid argument with existing users who have managed to work around it, but invalid with respect to new users of the function for whom it will be unreliable, e.g. the original poster of the underlying issue (not me).
> 
> Can I convert the PR to indicate a breaking change?

can you add a whatsnew note that shows the previous and new behavior; then it will be easier to assess

**Comment by attack68** [comment]
rendered whatsnew:
Screen Shot 2021-09-18 at 13 56 52

**Comment by attack68** [comment]
@jorisvandenbossche perhaps you care to review the whatsnew above?

**Comment by attack68** [comment]
noting #43840

**Comment by attack68** [comment]
linking #45475 and #45434 for any relevance if anyone takes this forward

**Comment by mroeschke** [comment]
Appears that this PR has sufficiently stalled, so closing for now. Looks almost there is anyone wants to pick it up.

**Comment by jreback** [comment]
i think would either do
- remove .argsort entirely
- have it return a numpy array of indexers

it does the wrong thing now - while your PR does seems to fix it up - its essentially duplicating the sorting methods intent

**Comment by simonjayhawkins** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**